POST JSON Rule definition can result in NPE

```
$ https -v --json POST :8181/api/v2/rules name:foo matchExpression="target.annotations.cryostat['JAVA_MAIN']=='demoApp'" eventSpecifier="template=Continuous,type=TARGET"
POST /api/v2/rules HTTP/1.1
Accept: application/json, */*;q=0.5
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 127
Content-Type: application/json
Host: localhost:8181
User-Agent: HTTPie/2.4.0
name: foo

{
    "eventSpecifier": "template=Continuous,type=TARGET",
    "matchExpression": "target.annotations.cryostat['JAVA_MAIN']=='demoApp'"
}


HTTP/1.1 500 Internal Server Error
content-encoding: gzip
content-length: 103
content-type: application/json

{
    "data": {
        "reason": null
    },
    "meta": {
        "status": "Internal Server Error",
        "type": "text/plain"
    }
}
```

I sent a `POST /api/v2/rules` and incorrectly formatted my request, so the `name` key was not included in the JSON body (instead, it was sent as an HTTP header). This resulted in an HTTP 500:

```
Aug 12, 2021 7:28:28 PM io.cryostat.core.log.Logger error
SEVERE: Exception thrown
io.cryostat.net.web.http.api.v2.ApiException: Internal Server Error
Caused by: java.lang.NullPointerException
	at io.cryostat.rules.Rule$Builder.from(Rule.java:234)
	at io.cryostat.util.RuleDeserializer.deserialize(RuleDeserializer.java:58)
	at io.cryostat.util.RuleDeserializer.deserialize(RuleDeserializer.java:51)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.google.gson.Gson.fromJson(Gson.java:846)
	at com.google.gson.Gson.fromJson(Gson.java:817)
	at io.cryostat.net.web.http.api.v2.RulesPostHandler.handle(RulesPostHandler.java:144)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:114)
	at io.cryostat.net.web.http.api.v2.AbstractV2RequestHandler.handle(AbstractV2RequestHandler.java:68)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)

Aug 12, 2021 7:28:28 PM io.cryostat.core.log.Logger info
INFO: (10.0.2.100:60442): POST /api/v2/rules 500 4ms
```

This should have instead been a `400`.

--- a/HTTP_API.md
+++ b/HTTP_API.md
@@ -1398,8 +1398,9 @@ The handler-specific descriptions below describe how each handler populates the
     attributes `"name"`, `"matchExpression"`, and `"eventSpecifier"` must be
     provided.
 
-    `"name"`: the name of this rule definition. This must be unique. This name
-    will also be used to generate the name of the associated recordings.
+    `"name"`: the name of this rule definition. This must be unique, except in
+    the case of "archiver rules" (see `eventSpecifier` below). This name will
+    also be used to generate the name of the associated recordings.
 
     `"matchExpression"`: a string expression used to determine which target JVMs
     this rule will apply to. The expression has a variable named `target` in
@@ -1412,9 +1413,15 @@ The handler-specific descriptions below describe how each handler populates the
     The simple expression `true` may also be used to create a rule which applies
     to any and all discovered targets.
 
-    `"eventSpecifier"`: a string of the form `template=Foo,type=TYPE`. This
+    `"eventSpecifier"`: a string of the form `template=Foo,type=TYPE`, which
     defines the event template that will be used for creating new recordings in
-    matching targets.
+    matching targets; or, the special string `"archive"`, which signifies that
+    this rule should cause all matching targets to have their current (at the
+    time of rule creation) JFR data copied to the Cryostat archives as a
+    one-time operation. When using `"archive"`, it is invalid to provide
+    `archivalPeriodSeconds`, `preservedArchives`, `maxSizeBytes`, or
+    `maxAgeSeconds`. Such "archiver rules" are only processed once and are not
+    persisted, so the `name` and `description` become optional.
 
     The following attributes are optional:
 

--- a/src/main/java/io/cryostat/rules/MatchExpressionValidator.java
+++ b/src/main/java/io/cryostat/rules/MatchExpressionValidator.java
@@ -49,7 +49,11 @@ public class MatchExpressionValidator {
 
     String validate(Rule rule) throws MatchExpressionValidationException {
         try {
-            CompilationUnitTree cut = parser.parse(rule.getName(), rule.getMatchExpression(), null);
+            String name = rule.getName();
+            if (name == null) {
+                name = "";
+            }
+            CompilationUnitTree cut = parser.parse(name, rule.getMatchExpression(), null);
             if (cut == null) {
                 throw new IllegalMatchExpressionException();
             }

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -64,10 +64,14 @@ public class Rule {
     private final int maxSizeBytes;
 
     Rule(Builder builder) throws MatchExpressionValidationException {
-        this.name = sanitizeRuleName(requireNonBlank(builder.name, Attribute.NAME));
+        this.eventSpecifier = builder.eventSpecifier;
+        if (isArchiver()) {
+            this.name = builder.name;
+        } else {
+            this.name = sanitizeRuleName(requireNonBlank(builder.name, Attribute.NAME));
+        }
         this.description = builder.description == null ? "" : builder.description;
         this.matchExpression = builder.matchExpression;
-        this.eventSpecifier = builder.eventSpecifier;
         this.archivalPeriodSeconds = builder.archivalPeriodSeconds;
         this.preservedArchives = builder.preservedArchives;
         this.maxAgeSeconds =
@@ -131,11 +135,8 @@ public class Rule {
     }
 
     public void validate() throws IllegalArgumentException, MatchExpressionValidationException {
-        requireNonBlank(this.name, Attribute.NAME);
         requireNonBlank(this.matchExpression, Attribute.MATCH_EXPRESSION);
         validateEventSpecifier(requireNonBlank(this.eventSpecifier, Attribute.EVENT_SPECIFIER));
-        requireNonNegative(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
-        requireNonNegative(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
         validateMatchExpression(this);
 
         if (isArchiver()) {
@@ -143,6 +144,10 @@ public class Rule {
             requireNonPositive(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
             requireNonPositive(this.maxSizeBytes, Attribute.MAX_SIZE_BYTES);
             requireNonPositive(this.maxAgeSeconds, Attribute.MAX_AGE_SECONDS);
+        } else {
+            requireNonBlank(this.name, Attribute.NAME);
+            requireNonNegative(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
+            requireNonNegative(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
         }
     }
 
@@ -188,10 +193,10 @@ public class Rule {
     }
 
     public static class Builder {
-        private String name;
-        private String description;
-        private String matchExpression;
-        private String eventSpecifier;
+        private String name = "";
+        private String description = "";
+        private String matchExpression = "";
+        private String eventSpecifier = "";
         private int archivalPeriodSeconds = 0;
         private int preservedArchives = 0;
         private int maxAgeSeconds = -1;

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -39,6 +39,8 @@ package io.cryostat.rules;
 
 import java.util.function.Function;
 
+import io.cryostat.recordings.RecordingTargetHelper;
+
 import com.google.gson.JsonObject;
 import io.vertx.core.MultiMap;
 import org.apache.commons.lang3.StringUtils;
@@ -114,16 +116,26 @@ public class Rule {
         return name.replaceAll("\\s", "_");
     }
 
-    static String validateMatchExpression(Rule rule) throws MatchExpressionValidationException {
-        return MATCH_EXPRESSION_VALIDATOR.validate(rule);
-    }
-
     private static String requireNonBlank(String s, Attribute attr) {
         if (StringUtils.isBlank(s)) {
             throw new IllegalArgumentException(
                     String.format("\"%s\" cannot be blank, was \"%s\"", attr, s));
         }
         return s;
+    }
+
+    public void validate() throws IllegalArgumentException, MatchExpressionValidationException {
+        requireNonBlank(this.name, Attribute.NAME);
+        requireNonBlank(this.matchExpression, Attribute.MATCH_EXPRESSION);
+        validateEventSpecifier(requireNonBlank(this.eventSpecifier, Attribute.EVENT_SPECIFIER));
+        requireNonNegative(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
+        requireNonNegative(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
+        validateMatchExpression(this);
+    }
+
+    private static String validateMatchExpression(Rule rule)
+            throws MatchExpressionValidationException {
+        return MATCH_EXPRESSION_VALIDATOR.validate(rule);
     }
 
     private static int requireNonNegative(int i, Attribute attr) {
@@ -134,13 +146,11 @@ public class Rule {
         return i;
     }
 
-    public void validate() throws IllegalArgumentException, MatchExpressionValidationException {
-        requireNonBlank(this.name, Attribute.NAME);
-        requireNonBlank(this.matchExpression, Attribute.MATCH_EXPRESSION);
-        requireNonBlank(this.eventSpecifier, Attribute.EVENT_SPECIFIER);
-        requireNonNegative(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
-        requireNonNegative(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
-        validateMatchExpression(this);
+    private static String validateEventSpecifier(String eventSpecifier)
+            throws IllegalArgumentException {
+        // throws if cannot be parsed
+        RecordingTargetHelper.parseEventSpecifierToTemplate(eventSpecifier);
+        return eventSpecifier;
     }
 
     @Override

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -39,13 +39,15 @@ package io.cryostat.rules;
 
 import java.util.function.Function;
 
-import io.cryostat.recordings.RecordingTargetHelper;
-
+import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
-import io.vertx.core.MultiMap;
+
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.lang3.builder.EqualsBuilder;
 import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import io.cryostat.recordings.RecordingTargetHelper;
+import io.vertx.core.MultiMap;
 
 public class Rule {
 
@@ -270,13 +272,11 @@ public class Rule {
         public static Builder from(JsonObject jsonObj) throws IllegalArgumentException {
             Rule.Builder builder =
                     new Rule.Builder()
-                            .name(jsonObj.get(Rule.Attribute.NAME.getSerialKey()).getAsString())
+                            .name(getAsNullableString(jsonObj, Rule.Attribute.NAME))
                             .matchExpression(
                                     jsonObj.get(Rule.Attribute.MATCH_EXPRESSION.getSerialKey())
                                             .getAsString())
-                            .description(
-                                    jsonObj.get(Rule.Attribute.DESCRIPTION.getSerialKey())
-                                            .getAsString())
+                            .description(getAsNullableString(jsonObj, Rule.Attribute.DESCRIPTION))
                             .eventSpecifier(
                                     jsonObj.get(Rule.Attribute.EVENT_SPECIFIER.getSerialKey())
                                             .getAsString());
@@ -286,6 +286,14 @@ public class Rule {
             builder.setOptionalInt(Rule.Attribute.MAX_SIZE_BYTES, jsonObj);
 
             return builder;
+        }
+
+        private static String getAsNullableString(JsonObject jsonObj, Rule.Attribute attr) {
+            JsonElement el = jsonObj.get(attr.getSerialKey());
+            if (el == null) {
+                return null;
+            }
+            return el.getAsString();
         }
 
         private Builder setOptionalInt(Rule.Attribute key, MultiMap formAttributes)

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -39,15 +39,14 @@ package io.cryostat.rules;
 
 import java.util.function.Function;
 
+import io.cryostat.recordings.RecordingTargetHelper;
+
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
-
+import io.vertx.core.MultiMap;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.lang3.builder.EqualsBuilder;
 import org.apache.commons.lang3.builder.HashCodeBuilder;
-
-import io.cryostat.recordings.RecordingTargetHelper;
-import io.vertx.core.MultiMap;
 
 public class Rule {
 

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -52,6 +52,8 @@ public class Rule {
     private static final MatchExpressionValidator MATCH_EXPRESSION_VALIDATOR =
             new MatchExpressionValidator();
 
+    static final String ARCHIVE_EVENT = "archive";
+
     private final String name;
     private final String description;
     private final String matchExpression;
@@ -93,6 +95,10 @@ public class Rule {
 
     public String getEventSpecifier() {
         return this.eventSpecifier;
+    }
+
+    public boolean isArchiver() {
+        return ARCHIVE_EVENT.equals(getEventSpecifier());
     }
 
     public int getArchivalPeriodSeconds() {
@@ -148,6 +154,9 @@ public class Rule {
 
     private static String validateEventSpecifier(String eventSpecifier)
             throws IllegalArgumentException {
+        if (eventSpecifier.equals(ARCHIVE_EVENT)) {
+            return eventSpecifier;
+        }
         // throws if cannot be parsed
         RecordingTargetHelper.parseEventSpecifierToTemplate(eventSpecifier);
         return eventSpecifier;

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -177,8 +177,8 @@ public class Rule {
         private String description;
         private String matchExpression;
         private String eventSpecifier;
-        private int archivalPeriodSeconds = 30;
-        private int preservedArchives = 1;
+        private int archivalPeriodSeconds = 0;
+        private int preservedArchives = 0;
         private int maxAgeSeconds = -1;
         private int maxSizeBytes = -1;
 

--- a/src/main/java/io/cryostat/rules/Rule.java
+++ b/src/main/java/io/cryostat/rules/Rule.java
@@ -137,6 +137,13 @@ public class Rule {
         requireNonNegative(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
         requireNonNegative(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
         validateMatchExpression(this);
+
+        if (isArchiver()) {
+            requireNonPositive(this.archivalPeriodSeconds, Attribute.ARCHIVAL_PERIOD_SECONDS);
+            requireNonPositive(this.preservedArchives, Attribute.PRESERVED_ARCHIVES);
+            requireNonPositive(this.maxSizeBytes, Attribute.MAX_SIZE_BYTES);
+            requireNonPositive(this.maxAgeSeconds, Attribute.MAX_AGE_SECONDS);
+        }
     }
 
     private static String validateMatchExpression(Rule rule)
@@ -148,6 +155,14 @@ public class Rule {
         if (i < 0) {
             throw new IllegalArgumentException(
                     String.format("\"%s\" cannot be negative, was \"%d\"", attr, i));
+        }
+        return i;
+    }
+
+    private static int requireNonPositive(int i, Attribute attr) {
+        if (i > 0) {
+            throw new IllegalArgumentException(
+                    String.format("\"%s\" cannot be positive, was \"%d\"", attr, i));
         }
         return i;
     }

--- a/src/main/java/io/cryostat/rules/RuleProcessor.java
+++ b/src/main/java/io/cryostat/rules/RuleProcessor.java
@@ -255,13 +255,12 @@ public class RuleProcessor
                     RecordingOptionsBuilder builder =
                             recordingOptionsBuilderFactory
                                     .create(connection.getService())
-                                    .name(rule.getRecordingName())
-                                    .toDisk(true);
+                                    .name(rule.getRecordingName());
                     if (rule.getMaxAgeSeconds() > 0) {
-                        builder = builder.maxAge(rule.getMaxAgeSeconds());
+                        builder = builder.maxAge(rule.getMaxAgeSeconds()).toDisk(true);
                     }
                     if (rule.getMaxSizeBytes() > 0) {
-                        builder = builder.maxSize(rule.getMaxSizeBytes());
+                        builder = builder.maxSize(rule.getMaxSizeBytes()).toDisk(true);
                     }
                     Pair<String, TemplateType> template =
                             RecordingTargetHelper.parseEventSpecifierToTemplate(

--- a/src/main/java/io/cryostat/rules/RuleProcessor.java
+++ b/src/main/java/io/cryostat/rules/RuleProcessor.java
@@ -47,6 +47,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 
 import org.openjdk.jmc.flightrecorder.configuration.recording.RecordingOptionsBuilder;
+import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
 
 import io.cryostat.configuration.CredentialsManager;
 import io.cryostat.core.log.Logger;
@@ -159,29 +160,38 @@ public class RuleProcessor
 
         Credentials credentials =
                 credentialsManager.getCredentials(serviceRef.getServiceUri().toString());
-        try {
-            startRuleRecording(new ConnectionDescriptor(serviceRef, credentials), rule);
-        } catch (Exception e) {
-            logger.error(e);
-        }
-
         logger.trace("Rule activation successful");
-        if (rule.getPreservedArchives() <= 0 || rule.getArchivalPeriodSeconds() <= 0) {
-            return;
+
+        if (rule.isArchiver()) {
+            try {
+                archiveRuleRecording(new ConnectionDescriptor(serviceRef, credentials), rule);
+            } catch (Exception e) {
+                logger.error(e);
+            }
+        } else {
+            try {
+                startRuleRecording(new ConnectionDescriptor(serviceRef, credentials), rule);
+            } catch (Exception e) {
+                logger.error(e);
+            }
+
+            if (rule.getPreservedArchives() <= 0 || rule.getArchivalPeriodSeconds() <= 0) {
+                return;
+            }
+            tasks.put(
+                    Pair.of(serviceRef, rule),
+                    scheduler.scheduleAtFixedRate(
+                            periodicArchiverFactory.create(
+                                    serviceRef,
+                                    credentialsManager,
+                                    rule,
+                                    recordingArchiveHelper,
+                                    this::archivalFailureHandler,
+                                    base32),
+                            rule.getArchivalPeriodSeconds(),
+                            rule.getArchivalPeriodSeconds(),
+                            TimeUnit.SECONDS));
         }
-        tasks.put(
-                Pair.of(serviceRef, rule),
-                scheduler.scheduleAtFixedRate(
-                        periodicArchiverFactory.create(
-                                serviceRef,
-                                credentialsManager,
-                                rule,
-                                recordingArchiveHelper,
-                                this::archivalFailureHandler,
-                                base32),
-                        rule.getArchivalPeriodSeconds(),
-                        rule.getArchivalPeriodSeconds(),
-                        TimeUnit.SECONDS));
     }
 
     private void deactivate(Rule rule, ServiceRef serviceRef) {
@@ -215,6 +225,25 @@ public class RuleProcessor
             task.cancel(true);
         }
         return null;
+    }
+
+    private void archiveRuleRecording(ConnectionDescriptor connectionDescriptor, Rule rule)
+            throws Exception {
+        targetConnectionManager.executeConnectedTask(
+                connectionDescriptor,
+                connection -> {
+                    IRecordingDescriptor descriptor =
+                            connection.getService().getSnapshotRecording();
+                    try {
+                        recordingArchiveHelper
+                                .saveRecording(connectionDescriptor, descriptor.getName())
+                                .get();
+                    } finally {
+                        connection.getService().close(descriptor);
+                    }
+
+                    return null;
+                });
     }
 
     private void startRuleRecording(ConnectionDescriptor connectionDescriptor, Rule rule)

--- a/src/main/java/io/cryostat/rules/RuleProcessor.java
+++ b/src/main/java/io/cryostat/rules/RuleProcessor.java
@@ -235,7 +235,7 @@ public class RuleProcessor
                         builder = builder.maxSize(rule.getMaxSizeBytes());
                     }
                     Pair<String, TemplateType> template =
-                            recordingTargetHelper.parseEventSpecifierToTemplate(
+                            RecordingTargetHelper.parseEventSpecifierToTemplate(
                                     rule.getEventSpecifier());
                     recordingTargetHelper.startRecording(
                             true,

--- a/src/main/java/io/cryostat/rules/RuleRegistry.java
+++ b/src/main/java/io/cryostat/rules/RuleRegistry.java
@@ -100,14 +100,16 @@ public class RuleRegistry extends AbstractEventEmitter<RuleEvent, Rule> {
                             "Rule with name \"%s\" already exists; refusing to overwrite",
                             rule.getName()));
         }
-        Path destination = rulesDir.resolve(rule.getName() + ".json");
-        this.fs.writeString(
-                destination,
-                gson.toJson(rule),
-                StandardOpenOption.WRITE,
-                StandardOpenOption.CREATE,
-                StandardOpenOption.TRUNCATE_EXISTING);
-        loadRules();
+        if (!rule.isArchiver()) {
+            Path destination = rulesDir.resolve(rule.getName() + ".json");
+            this.fs.writeString(
+                    destination,
+                    gson.toJson(rule),
+                    StandardOpenOption.WRITE,
+                    StandardOpenOption.CREATE,
+                    StandardOpenOption.TRUNCATE_EXISTING);
+            loadRules();
+        }
         emit(RuleEvent.ADDED, rule);
         return rule;
     }
@@ -138,7 +140,10 @@ public class RuleRegistry extends AbstractEventEmitter<RuleEvent, Rule> {
         if (!serviceRef.getAlias().isPresent()) {
             return Set.of();
         }
-        return rules.stream().filter(r -> applies(r, serviceRef)).collect(Collectors.toSet());
+        return rules.stream()
+                .filter(r -> !r.isArchiver())
+                .filter(r -> applies(r, serviceRef))
+                .collect(Collectors.toSet());
     }
 
     public Set<Rule> getRules() {

--- a/src/main/java/io/cryostat/rules/RuleRegistry.java
+++ b/src/main/java/io/cryostat/rules/RuleRegistry.java
@@ -94,13 +94,13 @@ public class RuleRegistry extends AbstractEventEmitter<RuleEvent, Rule> {
     }
 
     public Rule addRule(Rule rule) throws IOException {
-        if (hasRuleByName(rule.getName())) {
-            throw new RuleException(
-                    String.format(
-                            "Rule with name \"%s\" already exists; refusing to overwrite",
-                            rule.getName()));
-        }
         if (!rule.isArchiver()) {
+            if (hasRuleByName(rule.getName())) {
+                throw new RuleException(
+                        String.format(
+                                "Rule with name \"%s\" already exists; refusing to overwrite",
+                                rule.getName()));
+            }
             Path destination = rulesDir.resolve(rule.getName() + ".json");
             this.fs.writeString(
                     destination,

--- a/src/main/java/io/cryostat/rules/RuleRegistry.java
+++ b/src/main/java/io/cryostat/rules/RuleRegistry.java
@@ -140,9 +140,7 @@ public class RuleRegistry extends AbstractEventEmitter<RuleEvent, Rule> {
         if (!serviceRef.getAlias().isPresent()) {
             return Set.of();
         }
-        return rules.stream()
-                .filter(r -> applies(r, serviceRef))
-                .collect(Collectors.toSet());
+        return rules.stream().filter(r -> applies(r, serviceRef)).collect(Collectors.toSet());
     }
 
     public Set<Rule> getRules() {

--- a/src/main/java/io/cryostat/rules/RuleRegistry.java
+++ b/src/main/java/io/cryostat/rules/RuleRegistry.java
@@ -141,7 +141,6 @@ public class RuleRegistry extends AbstractEventEmitter<RuleEvent, Rule> {
             return Set.of();
         }
         return rules.stream()
-                .filter(r -> !r.isArchiver())
                 .filter(r -> applies(r, serviceRef))
                 .collect(Collectors.toSet());
     }

--- a/src/test/java/io/cryostat/rules/RuleProcessorTest.java
+++ b/src/test/java/io/cryostat/rules/RuleProcessorTest.java
@@ -39,6 +39,7 @@ package io.cryostat.rules;
 
 import java.net.URI;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
@@ -47,6 +48,7 @@ import java.util.function.Function;
 import org.openjdk.jmc.common.unit.IConstrainedMap;
 import org.openjdk.jmc.flightrecorder.configuration.recording.RecordingOptionsBuilder;
 import org.openjdk.jmc.rjmx.services.jfr.IFlightRecorderService;
+import org.openjdk.jmc.rjmx.services.jfr.IRecordingDescriptor;
 
 import io.cryostat.configuration.CredentialsManager;
 import io.cryostat.core.log.Logger;
@@ -72,6 +74,7 @@ import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
 import org.mockito.Mock;
 import org.mockito.Mockito;
 import org.mockito.junit.jupiter.MockitoExtension;
@@ -232,6 +235,64 @@ class RuleProcessorTest {
         MatcherAssert.assertThat(templateTypeCaptor.getValue(), Matchers.nullValue());
 
         Mockito.verify(scheduler).scheduleAtFixedRate(periodicArchiver, 67, 67, TimeUnit.SECONDS);
+    }
+
+    @Test
+    void testSuccessfulArchiverRuleActivationWithCredentials() throws Exception {
+        Mockito.when(targetConnectionManager.executeConnectedTask(Mockito.any(), Mockito.any()))
+                .thenAnswer(
+                        arg0 ->
+                                ((TargetConnectionManager.ConnectedTask<Object>)
+                                                arg0.getArgument(1))
+                                        .execute(connection));
+        Mockito.when(connection.getService()).thenReturn(service);
+
+        IRecordingDescriptor snapshot = Mockito.mock(IRecordingDescriptor.class);
+        Mockito.when(snapshot.getName()).thenReturn("Snapshot-1");
+        Mockito.when(service.getSnapshotRecording()).thenReturn(snapshot);
+
+        String jmxUrl = "service:jmx:rmi://localhost:9091/jndi/rmi://fooHost:9091/jmxrmi";
+        ServiceRef serviceRef = new ServiceRef(new URI(jmxUrl), "com.example.App");
+
+        Credentials credentials = new Credentials("foouser", "barpassword");
+        Mockito.when(credentialsManager.getCredentials(jmxUrl)).thenReturn(credentials);
+
+        TargetDiscoveryEvent tde = new TargetDiscoveryEvent(EventKind.FOUND, serviceRef);
+
+        Rule rule =
+                new Rule.Builder()
+                        .name("Test Rule")
+                        .description("Automated unit test rule")
+                        .matchExpression("target.alias == 'com.example.App'")
+                        .eventSpecifier("archive")
+                        .build();
+
+        Mockito.when(registry.getRules(serviceRef)).thenReturn(Set.of(rule));
+
+        Mockito.when(recordingArchiveHelper.saveRecording(Mockito.any(), Mockito.any()))
+                .thenReturn(CompletableFuture.completedFuture("unusedPath"));
+
+        processor.accept(tde);
+
+        ArgumentCaptor<ConnectionDescriptor> connectionDescriptorCaptor =
+                ArgumentCaptor.forClass(ConnectionDescriptor.class);
+        ArgumentCaptor<String> recordingSaveNameCaptor = ArgumentCaptor.forClass(String.class);
+
+        InOrder inOrder = Mockito.inOrder(service, recordingArchiveHelper);
+        inOrder.verify(service).getSnapshotRecording();
+        inOrder.verify(recordingArchiveHelper)
+                .saveRecording(
+                        connectionDescriptorCaptor.capture(), recordingSaveNameCaptor.capture());
+        inOrder.verify(service).close(snapshot);
+
+        ConnectionDescriptor connectionDescriptor = connectionDescriptorCaptor.getValue();
+        MatcherAssert.assertThat(
+                connectionDescriptor.getTargetId(),
+                Matchers.equalTo(serviceRef.getServiceUri().toString()));
+        MatcherAssert.assertThat(
+                connectionDescriptor.getCredentials().get(), Matchers.equalTo(credentials));
+        MatcherAssert.assertThat(
+                recordingSaveNameCaptor.getValue(), Matchers.equalTo(snapshot.getName()));
     }
 
     @Test

--- a/src/test/java/io/cryostat/rules/RuleRegistryTest.java
+++ b/src/test/java/io/cryostat/rules/RuleRegistryTest.java
@@ -45,11 +45,15 @@ import java.nio.file.StandardOpenOption;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
 
 import io.cryostat.MainModule;
 import io.cryostat.core.log.Logger;
 import io.cryostat.core.sys.FileSystem;
 import io.cryostat.platform.ServiceRef;
+import io.cryostat.rules.RuleRegistry.RuleEvent;
+import io.cryostat.util.events.Event;
 
 import com.google.gson.Gson;
 import org.hamcrest.MatcherAssert;
@@ -128,6 +132,9 @@ class RuleRegistryTest {
         Mockito.when(fs.listDirectoryChildren(rulesDir)).thenReturn(List.of("test_rule.json"));
         Mockito.when(fs.readFile(rulePath)).thenReturn(fileReader);
 
+        CompletableFuture<Event<RuleEvent, Rule>> eventListener = new CompletableFuture<>();
+        registry.addListener(eventListener::complete);
+
         registry.addRule(testRule);
 
         InOrder inOrder = Mockito.inOrder(gson, fs);
@@ -141,6 +148,10 @@ class RuleRegistryTest {
                         StandardOpenOption.CREATE,
                         StandardOpenOption.TRUNCATE_EXISTING);
         inOrder.verify(fs).listDirectoryChildren(rulesDir);
+
+        Event<RuleEvent, Rule> event = eventListener.get(1, TimeUnit.SECONDS);
+        MatcherAssert.assertThat(event.getEventType(), Matchers.equalTo(RuleEvent.ADDED));
+        MatcherAssert.assertThat(event.getPayload(), Matchers.sameInstance(testRule));
     }
 
     @Test
@@ -169,6 +180,32 @@ class RuleRegistryTest {
         registry.loadRules();
 
         Assertions.assertThrows(IOException.class, () -> registry.addRule(testRule));
+    }
+
+    @Test
+    void testAddRuleAllowsDuplicateNameOnArchivers() throws Exception {
+        Path rulePath = Mockito.mock(Path.class);
+        Mockito.when(rulesDir.resolve(Mockito.anyString())).thenReturn(rulePath);
+        Mockito.when(fs.listDirectoryChildren(rulesDir)).thenReturn(List.of("test_rule.json"));
+        Mockito.when(fs.readFile(rulePath)).thenReturn(fileReader);
+
+        registry.loadRules();
+
+        Rule archiver =
+                new Rule.Builder()
+                        .name(testRule.getName())
+                        .matchExpression(testRule.getMatchExpression())
+                        .description(testRule.getDescription())
+                        .eventSpecifier("archive")
+                        .build();
+
+        CompletableFuture<Event<RuleEvent, Rule>> eventListener = new CompletableFuture<>();
+        registry.addListener(eventListener::complete);
+
+        Assertions.assertDoesNotThrow(() -> registry.addRule(archiver));
+        Event<RuleEvent, Rule> event = eventListener.get(1, TimeUnit.SECONDS);
+        MatcherAssert.assertThat(event.getEventType(), Matchers.equalTo(RuleEvent.ADDED));
+        MatcherAssert.assertThat(event.getPayload(), Matchers.sameInstance(archiver));
     }
 
     @Test
@@ -209,6 +246,23 @@ class RuleRegistryTest {
         MatcherAssert.assertThat(
                 registry.getRules(new ServiceRef(null, "com.example.App")),
                 Matchers.equalTo(Set.of(testRule)));
+    }
+
+    @Test
+    void testGetRulesByServiceRefIgnoresArchivers() throws Exception {
+        Rule archiverRule =
+                new Rule.Builder()
+                        .name(testRule.getName())
+                        .matchExpression(testRule.getMatchExpression())
+                        .description(testRule.getDescription())
+                        .eventSpecifier("archive")
+                        .build();
+
+        registry.addRule(archiverRule);
+
+        MatcherAssert.assertThat(
+                registry.getRules(new ServiceRef(null, "com.example.App")),
+                Matchers.equalTo(Set.of()));
     }
 
     @Test

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -209,8 +209,24 @@ class RuleTest {
 
     @Test
     void shouldAcceptEventSpecifierArchiveSpecialCase() throws Exception {
+        Rule rule = builder.matchExpression(MATCH_EXPRESSION).eventSpecifier("archive").build();
+        MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo("archive"));
+    }
+
+    @Test
+    void shouldAcceptEventSpecifierArchiveSpecialCaseWithName() throws Exception {
         Rule rule =
                 builder.name("Some Rule")
+                        .matchExpression(MATCH_EXPRESSION)
+                        .eventSpecifier("archive")
+                        .build();
+        MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo("archive"));
+    }
+
+    @Test
+    void shouldAcceptEventSpecifierArchiveSpecialCaseWithDescription() throws Exception {
+        Rule rule =
+                builder.description("Unused description")
                         .matchExpression(MATCH_EXPRESSION)
                         .eventSpecifier("archive")
                         .build();
@@ -223,8 +239,7 @@ class RuleTest {
                 Assertions.assertThrows(
                         IllegalArgumentException.class,
                         () -> {
-                            builder.name(NAME)
-                                    .matchExpression(MATCH_EXPRESSION)
+                            builder.matchExpression(MATCH_EXPRESSION)
                                     .eventSpecifier("archive")
                                     .archivalPeriodSeconds(5)
                                     .build();
@@ -240,8 +255,7 @@ class RuleTest {
                 Assertions.assertThrows(
                         IllegalArgumentException.class,
                         () -> {
-                            builder.name(NAME)
-                                    .matchExpression(MATCH_EXPRESSION)
+                            builder.matchExpression(MATCH_EXPRESSION)
                                     .eventSpecifier("archive")
                                     .preservedArchives(5)
                                     .build();
@@ -257,8 +271,7 @@ class RuleTest {
                 Assertions.assertThrows(
                         IllegalArgumentException.class,
                         () -> {
-                            builder.name(NAME)
-                                    .matchExpression(MATCH_EXPRESSION)
+                            builder.matchExpression(MATCH_EXPRESSION)
                                     .eventSpecifier("archive")
                                     .maxSizeBytes(5)
                                     .build();
@@ -274,8 +287,7 @@ class RuleTest {
                 Assertions.assertThrows(
                         IllegalArgumentException.class,
                         () -> {
-                            builder.name(NAME)
-                                    .matchExpression(MATCH_EXPRESSION)
+                            builder.matchExpression(MATCH_EXPRESSION)
                                     .eventSpecifier("archive")
                                     .maxAgeSeconds(5)
                                     .build();

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -216,4 +216,72 @@ class RuleTest {
                         .build();
         MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo("archive"));
     }
+
+    @Test
+    void shouldThrowOnArchiverWithArchivalPeriod() throws Exception {
+        IllegalArgumentException ex =
+                Assertions.assertThrows(
+                        IllegalArgumentException.class,
+                        () -> {
+                            builder.name(NAME)
+                                    .matchExpression(MATCH_EXPRESSION)
+                                    .eventSpecifier("archive")
+                                    .archivalPeriodSeconds(5)
+                                    .build();
+                        });
+        MatcherAssert.assertThat(
+                ex.getMessage(),
+                Matchers.containsString("\"archivalPeriodSeconds\" cannot be positive, was \"5\""));
+    }
+
+    @Test
+    void shouldThrowOnArchiverWithPreservedArchives() throws Exception {
+        IllegalArgumentException ex =
+                Assertions.assertThrows(
+                        IllegalArgumentException.class,
+                        () -> {
+                            builder.name(NAME)
+                                    .matchExpression(MATCH_EXPRESSION)
+                                    .eventSpecifier("archive")
+                                    .preservedArchives(5)
+                                    .build();
+                        });
+        MatcherAssert.assertThat(
+                ex.getMessage(),
+                Matchers.containsString("\"preservedArchives\" cannot be positive, was \"5\""));
+    }
+
+    @Test
+    void shouldThrowOnArchiverWithMaxSizeBytes() throws Exception {
+        IllegalArgumentException ex =
+                Assertions.assertThrows(
+                        IllegalArgumentException.class,
+                        () -> {
+                            builder.name(NAME)
+                                    .matchExpression(MATCH_EXPRESSION)
+                                    .eventSpecifier("archive")
+                                    .maxSizeBytes(5)
+                                    .build();
+                        });
+        MatcherAssert.assertThat(
+                ex.getMessage(),
+                Matchers.containsString("\"maxSizeBytes\" cannot be positive, was \"5\""));
+    }
+
+    @Test
+    void shouldThrowOnArchiverWithMaxAgeSeconds() throws Exception {
+        IllegalArgumentException ex =
+                Assertions.assertThrows(
+                        IllegalArgumentException.class,
+                        () -> {
+                            builder.name(NAME)
+                                    .matchExpression(MATCH_EXPRESSION)
+                                    .eventSpecifier("archive")
+                                    .maxAgeSeconds(5)
+                                    .build();
+                        });
+        MatcherAssert.assertThat(
+                ex.getMessage(),
+                Matchers.containsString("\"maxAgeSeconds\" cannot be positive, was \"5\""));
+    }
 }

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -206,4 +206,14 @@ class RuleTest {
                         .build();
         MatcherAssert.assertThat(rule.getRecordingName(), Matchers.equalTo("auto_Some_Rule"));
     }
+
+    @Test
+    void shouldAcceptEventSpecifierArchiveSpecialCase() throws Exception {
+        Rule rule =
+                builder.name("Some Rule")
+                        .matchExpression(MATCH_EXPRESSION)
+                        .eventSpecifier("archive")
+                        .build();
+        MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo("archive"));
+    }
 }

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -123,7 +123,8 @@ class RuleTest {
 
     @ParameterizedTest
     @NullAndEmptySource
-    void shouldThrowOnBlankEventSpecifier(String s) {
+    @ValueSource(strings = {"invalid", "events=incorrect", "tenplate=typo"})
+    void shouldThrowOnInvalidEventSpecifier(String s) {
         IllegalArgumentException ex =
                 Assertions.assertThrows(
                         IllegalArgumentException.class,
@@ -135,7 +136,10 @@ class RuleTest {
                         });
         MatcherAssert.assertThat(
                 ex.getMessage(),
-                Matchers.containsString("\"eventSpecifier\" cannot be blank, was \"" + s + "\""));
+                Matchers.either(
+                                Matchers.equalTo(
+                                        "\"eventSpecifier\" cannot be blank, was \"" + s + "\""))
+                        .or(Matchers.equalTo(s)));
     }
 
     @Test

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -38,7 +38,7 @@
 package io.cryostat.rules;
 
 import com.google.gson.JsonObject;
-
+import io.vertx.core.MultiMap;
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.Matchers;
 import org.junit.jupiter.api.Assertions;
@@ -50,8 +50,6 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.NullAndEmptySource;
 import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.junit.jupiter.MockitoExtension;
-
-import io.vertx.core.MultiMap;
 
 @ExtendWith(MockitoExtension.class)
 class RuleTest {
@@ -309,7 +307,8 @@ class RuleTest {
         class Json {
 
             @Test
-            void testCompleteRule() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testCompleteRule()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String name = "Some Rule";
                 String description = "This is a description";
                 String matchExpression = "target.alias=='TheAlias'";
@@ -332,18 +331,21 @@ class RuleTest {
 
                 MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
                 MatcherAssert.assertThat(rule.getDescription(), Matchers.equalTo(description));
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
                 MatcherAssert.assertThat(rule.getMaxAgeSeconds(), Matchers.equalTo(maxAgeSeconds));
                 MatcherAssert.assertThat(rule.getMaxSizeBytes(), Matchers.equalTo(maxSizeBytes));
-                MatcherAssert.assertThat(rule.getArchivalPeriodSeconds(),
-                        Matchers.equalTo(archivalPeriodSeconds));
-                MatcherAssert.assertThat(rule.getPreservedArchives(),
-                        Matchers.equalTo(preservedArchives));
+                MatcherAssert.assertThat(
+                        rule.getArchivalPeriodSeconds(), Matchers.equalTo(archivalPeriodSeconds));
+                MatcherAssert.assertThat(
+                        rule.getPreservedArchives(), Matchers.equalTo(preservedArchives));
             }
 
             @Test
-            void testRuleWithoutOptionalFields() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testRuleWithoutOptionalFields()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String name = "Some Rule";
                 String matchExpression = "target.alias=='TheAlias'";
                 String eventSpecifier = "template=Foo";
@@ -355,12 +357,15 @@ class RuleTest {
                 Rule rule = Rule.Builder.from(json).build();
 
                 MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
             }
 
             @Test
-            void testArchiverWithoutName() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testArchiverWithoutName()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String matchExpression = "target.alias=='TheAlias'";
                 String eventSpecifier = "archive";
 
@@ -369,16 +374,19 @@ class RuleTest {
                 json.addProperty("eventSpecifier", eventSpecifier);
                 Rule rule = Rule.Builder.from(json).build();
 
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
             }
-
         }
+
         @Nested
         class Form {
 
             @Test
-            void testCompleteRule() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testCompleteRule()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String name = "Some Rule";
                 String description = "This is a description";
                 String matchExpression = "target.alias=='TheAlias'";
@@ -401,18 +409,21 @@ class RuleTest {
 
                 MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
                 MatcherAssert.assertThat(rule.getDescription(), Matchers.equalTo(description));
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
                 MatcherAssert.assertThat(rule.getMaxAgeSeconds(), Matchers.equalTo(maxAgeSeconds));
                 MatcherAssert.assertThat(rule.getMaxSizeBytes(), Matchers.equalTo(maxSizeBytes));
-                MatcherAssert.assertThat(rule.getArchivalPeriodSeconds(),
-                        Matchers.equalTo(archivalPeriodSeconds));
-                MatcherAssert.assertThat(rule.getPreservedArchives(),
-                        Matchers.equalTo(preservedArchives));
+                MatcherAssert.assertThat(
+                        rule.getArchivalPeriodSeconds(), Matchers.equalTo(archivalPeriodSeconds));
+                MatcherAssert.assertThat(
+                        rule.getPreservedArchives(), Matchers.equalTo(preservedArchives));
             }
 
             @Test
-            void testRuleWithoutOptionalFields() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testRuleWithoutOptionalFields()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String name = "Some Rule";
                 String matchExpression = "target.alias=='TheAlias'";
                 String eventSpecifier = "template=Foo";
@@ -424,12 +435,15 @@ class RuleTest {
                 Rule rule = Rule.Builder.from(form).build();
 
                 MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
             }
 
             @Test
-            void testArchiverWithoutName() throws IllegalArgumentException, MatchExpressionValidationException {
+            void testArchiverWithoutName()
+                    throws IllegalArgumentException, MatchExpressionValidationException {
                 String matchExpression = "target.alias=='TheAlias'";
                 String eventSpecifier = "archive";
 
@@ -438,11 +452,11 @@ class RuleTest {
                 form.set("eventSpecifier", eventSpecifier);
                 Rule rule = Rule.Builder.from(form).build();
 
-                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
-                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(
+                        rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(
+                        rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
             }
-
         }
-
     }
 }

--- a/src/test/java/io/cryostat/rules/RuleTest.java
+++ b/src/test/java/io/cryostat/rules/RuleTest.java
@@ -37,16 +37,21 @@
  */
 package io.cryostat.rules;
 
+import com.google.gson.JsonObject;
+
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.Matchers;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.NullAndEmptySource;
 import org.junit.jupiter.params.provider.ValueSource;
 import org.mockito.junit.jupiter.MockitoExtension;
+
+import io.vertx.core.MultiMap;
 
 @ExtendWith(MockitoExtension.class)
 class RuleTest {
@@ -295,5 +300,149 @@ class RuleTest {
         MatcherAssert.assertThat(
                 ex.getMessage(),
                 Matchers.containsString("\"maxAgeSeconds\" cannot be positive, was \"5\""));
+    }
+
+    @Nested
+    class Deserialization {
+
+        @Nested
+        class Json {
+
+            @Test
+            void testCompleteRule() throws IllegalArgumentException, MatchExpressionValidationException {
+                String name = "Some Rule";
+                String description = "This is a description";
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "template=Foo";
+                int maxAgeSeconds = 60;
+                int maxSizeBytes = 32 * 1024;
+                int archivalPeriodSeconds = 300;
+                int preservedArchives = 5;
+
+                JsonObject json = new JsonObject();
+                json.addProperty("name", name);
+                json.addProperty("description", description);
+                json.addProperty("matchExpression", matchExpression);
+                json.addProperty("eventSpecifier", eventSpecifier);
+                json.addProperty("maxAgeSeconds", maxAgeSeconds);
+                json.addProperty("maxSizeBytes", maxSizeBytes);
+                json.addProperty("archivalPeriodSeconds", archivalPeriodSeconds);
+                json.addProperty("preservedArchives", preservedArchives);
+                Rule rule = Rule.Builder.from(json).build();
+
+                MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
+                MatcherAssert.assertThat(rule.getDescription(), Matchers.equalTo(description));
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(rule.getMaxAgeSeconds(), Matchers.equalTo(maxAgeSeconds));
+                MatcherAssert.assertThat(rule.getMaxSizeBytes(), Matchers.equalTo(maxSizeBytes));
+                MatcherAssert.assertThat(rule.getArchivalPeriodSeconds(),
+                        Matchers.equalTo(archivalPeriodSeconds));
+                MatcherAssert.assertThat(rule.getPreservedArchives(),
+                        Matchers.equalTo(preservedArchives));
+            }
+
+            @Test
+            void testRuleWithoutOptionalFields() throws IllegalArgumentException, MatchExpressionValidationException {
+                String name = "Some Rule";
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "template=Foo";
+
+                JsonObject json = new JsonObject();
+                json.addProperty("name", name);
+                json.addProperty("matchExpression", matchExpression);
+                json.addProperty("eventSpecifier", eventSpecifier);
+                Rule rule = Rule.Builder.from(json).build();
+
+                MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+            }
+
+            @Test
+            void testArchiverWithoutName() throws IllegalArgumentException, MatchExpressionValidationException {
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "archive";
+
+                JsonObject json = new JsonObject();
+                json.addProperty("matchExpression", matchExpression);
+                json.addProperty("eventSpecifier", eventSpecifier);
+                Rule rule = Rule.Builder.from(json).build();
+
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+            }
+
+        }
+        @Nested
+        class Form {
+
+            @Test
+            void testCompleteRule() throws IllegalArgumentException, MatchExpressionValidationException {
+                String name = "Some Rule";
+                String description = "This is a description";
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "template=Foo";
+                int maxAgeSeconds = 60;
+                int maxSizeBytes = 32 * 1024;
+                int archivalPeriodSeconds = 300;
+                int preservedArchives = 5;
+
+                MultiMap form = MultiMap.caseInsensitiveMultiMap();
+                form.set("name", name);
+                form.set("description", description);
+                form.set("matchExpression", matchExpression);
+                form.set("eventSpecifier", eventSpecifier);
+                form.set("maxAgeSeconds", String.valueOf(maxAgeSeconds));
+                form.set("maxSizeBytes", String.valueOf(maxSizeBytes));
+                form.set("archivalPeriodSeconds", String.valueOf(archivalPeriodSeconds));
+                form.set("preservedArchives", String.valueOf(preservedArchives));
+                Rule rule = Rule.Builder.from(form).build();
+
+                MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
+                MatcherAssert.assertThat(rule.getDescription(), Matchers.equalTo(description));
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+                MatcherAssert.assertThat(rule.getMaxAgeSeconds(), Matchers.equalTo(maxAgeSeconds));
+                MatcherAssert.assertThat(rule.getMaxSizeBytes(), Matchers.equalTo(maxSizeBytes));
+                MatcherAssert.assertThat(rule.getArchivalPeriodSeconds(),
+                        Matchers.equalTo(archivalPeriodSeconds));
+                MatcherAssert.assertThat(rule.getPreservedArchives(),
+                        Matchers.equalTo(preservedArchives));
+            }
+
+            @Test
+            void testRuleWithoutOptionalFields() throws IllegalArgumentException, MatchExpressionValidationException {
+                String name = "Some Rule";
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "template=Foo";
+
+                MultiMap form = MultiMap.caseInsensitiveMultiMap();
+                form.set("name", name);
+                form.set("matchExpression", matchExpression);
+                form.set("eventSpecifier", eventSpecifier);
+                Rule rule = Rule.Builder.from(form).build();
+
+                MatcherAssert.assertThat(rule.getName(), Matchers.equalTo("Some_Rule"));
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+            }
+
+            @Test
+            void testArchiverWithoutName() throws IllegalArgumentException, MatchExpressionValidationException {
+                String matchExpression = "target.alias=='TheAlias'";
+                String eventSpecifier = "archive";
+
+                MultiMap form = MultiMap.caseInsensitiveMultiMap();
+                form.set("matchExpression", matchExpression);
+                form.set("eventSpecifier", eventSpecifier);
+                Rule rule = Rule.Builder.from(form).build();
+
+                MatcherAssert.assertThat(rule.getMatchExpression(), Matchers.equalTo(matchExpression));
+                MatcherAssert.assertThat(rule.getEventSpecifier(), Matchers.equalTo(eventSpecifier));
+            }
+
+        }
+
     }
 }

